### PR TITLE
IPV6 backends

### PR DESCRIPTION
**Pull request by krawthekrow**
Depends on #36.

### CODE REVIEW

**Comment by cryslith** [comment]
Now that #36 is merged, could you please rebase the relevant commits from this PR onto master?

**Comment by cryslith** [comment]
@ed1d1a8d [see above](https://github.com/sipb/spike/pull/38#issuecomment-335336308)

**Comment by krawthekrow** [comment]
Oops, missed that.

File: forward/rewriting.lua [hunk 1]
```diff
@@ -204,31 +204,42 @@ function Rewriting:process_packet(i, o)
...
+   else
+      error("backend length must be 4 (for IPv4) or 16 (for IPv6)")
+   end
+
    datagram:push(outer_ip_header)
 
    local outer_eth_header = Ethernet:new({src = self.src_mac or eth_dst,
                                           dst = self.dst_mac,
-                                          type = L3_IPV4})
+                                          type = l3_type})
```
Comment (cryslith) [inline forward/rewriting.lua]: This should be `outer_l3_type` not `l3_type`, right?

Comment (cryslith) [reply forward/rewriting.lua]: On a slightly different note: we should make sure to design our tests to catch issues like this.

**Review by cryslith — CHANGES_REQUESTED** [review]
lgtm (one potential bug)

**Comment by cryslith** [comment]
Oh, and please rebase this too!

File: forward/rewriting.lua (hunk 1 above)
Comment (krawthekrow) [reply forward/rewriting.lua]: Yeah, this was before I got the automated testing running. I'll fix the bug after adding a test to catch it.

**Comment by krawthekrow** [comment]
Is it true that we will always only have either all IPv4 backends or all IPv6 backends, not a mixture of both? Asking since the current implementation appears to assume that.

**Comment by cryslith** [comment]
@krawthekrow That is not a sound assumption.  I expect that we would have backends of both types.

**Comment by krawthekrow** [comment]
Yeah, thought so too. It would also be easier for testing if both were supported at once.

**Review by cryslith — APPROVED** [review]

**Comment by cryslith** [comment]
Superseded by #48.

**Pull request closed without merging**